### ISSUE TITLE
mca_fs_lustre.so depends on system hwloc even with --with-hwloc

### ISSUE THREAD

**Issue by benmenadue** [open-mpi/ompi#6349]
Hi,

This is for OpenMPI 4.0.0, using the release tarball.

Configuring with `--with-lustre` and `--with-hwloc` has given me this:
```
$ ldd ./lib/openmpi/mca_fs_lustre.so | grep '^\.\|hwloc'
./lib/openmpi/mca_fs_lustre.so:
        libhwloc.so.5 => /usr/lib64/libhwloc.so.5 (0x00007efc73b67000)
        libhwloc.so.15 => /apps/hwloc/2.0.3/lib/libhwloc.so.15 (0x00007efc7292b000)
```

It's picking up `libhwloc.so.5` from a direct dependency in that `.so`, and `libhwloc.so.15` through the dependency on `libmpi.so.40`:
```
$ objdump -p ./lib/openmpi/mca_fs_lustre.so | grep NEEDED | grep 'libmpi\.so\|hwloc'
  NEEDED               libmpi.so.40
  NEEDED               libhwloc.so.5
$ objdump -p ./lib/libmpi.so.40 | grep NEEDED | grep 'hwloc'
  NEEDED               libhwloc.so.15
```

Cheers,
Ben

**Comment by benmenadue** [comment]
Actually, this might be to do with how the `--with-lustre` option is handled... I'm seeing this:
```
configure:300583: checking for library containing llapi_file_create
configure:300614: gcc -std=gnu99 -std=gnu99 -o conftest -DNDEBUG -g -march=native -finline-functions -fno-strict-aliasing -pthread    -I/apps/hwloc/2.0.3/include  -I/apps/libevent/2.1.8/include    -L/apps/hwloc/2.0.3/lib  -L/apps/libevent/2.1.8/lib -L/usr/lib64 conftest.c -lrt -lm -lutil  -lz  -lhwloc  -levent -levent_pthreads >&5
```
It looks like that extra unnecessary `-L/usr/lib64` is being moved to the start during the relink, and so the `-lhwloc` is picking up the version there. Here the `relink_command` in the `.la` file:
```
relink_command="(cd /tmp/tmp.rGpZVljW8D/openmpi-4.0.0/gcc-opt/ompi/mca/fs/lustre; /bin/sh \"/tmp/tmp.rGpZVljW8D/openmpi-4.0.0/gcc-opt/libtool\"  --silent --tag CC --mode=relink gcc -std=gnu99 -std=gnu99 -DNDEBUG -g -march=native -finline-functions -fno-strict-aliasing -pthread -module -avoid-version -L/usr/lib64 -L/apps/hwloc/2.0.3/lib -L/apps/libevent/2.1.8/lib -o mca_fs_lustre.la -rpath /apps/openmpi-mofed4.2-pbs14.2/4.0.0/lib/openmpi fs_lustre.lo fs_lustre_component.lo fs_lustre_file_open.lo ../../../../ompi/libmpi.la -llustreapi -lrt -lm -lutil -lz -lhwloc -levent -levent_pthreads @inst_prefix_dir@)"
```

**Comment by benmenadue** [comment]
But that's not what's breaking my test, I don't think -- it looks like the application (Gromacs) itself is directly linked against the system hwloc, so that's being brought in before the one OpenMPI is dependent on. The hwloc symbols referenced from OpenMPI are then being bound to the system version of the library that Gromacs brought in, instead of the one OpenMPI brought in, and this is leading to a segfault.

Sorry for the noise, looks like it was my fault :) .

**Comment by bgoglin** [comment]
I confirm that Gromacs (at least optionally) uses hwloc internally.

**Comment by ggouaillardet** [comment]
@benmenadue I found some fishy things in Open MPI configury, `lib/openmpi/mca_fs_lustre.so` is obviously independant of any app such as Gromacs, so it should never be linked with two `hwloc` libraries.

**Comment by gpaulsen** [comment]
Open MPI now

**Comment by gpaulsen** [comment]
Oops preliminarily closed this after merged to `master`.  Keeping this issue open until the fix gets cherry-picked to `v5.0.x`

**Comment by awlauria** [comment]
Fixed by: https://github.com/open-mpi/ompi/pull/9121
